### PR TITLE
boost 1.70.0

### PR DESCRIPTION
**Pull request by fxcoudert**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by fxcoudert** [comment]
- `cgal` fails to build with:

```
-- Found Boost 1.70.0 at /usr/local/lib/cmake/Boost-1.70.0
--   Requested configuration: QUIET REQUIRED COMPONENTS thread;system
-- Found boost_headers 1.70.0 at /usr/local/lib/cmake/boost_headers-1.70.0
-- Found boost_thread 1.70.0 at /usr/local/lib/cmake/boost_thread-1.70.0
-- No suitable boost_thread variant has been identified!
--   libboost_thread-mt.dylib (shared, BUILD_SHARED_LIBS not ON, set Boost_USE_STATIC_LIBS=OFF to override)
--   libboost_thread-mt.a (static, Boost_USE_STATIC_LIBS=OFF)
CMake Error at /usr/local/lib/cmake/Boost-1.70.0/BoostConfig.cmake:95 (find_package):
  Found package configuration file:

    /usr/local/lib/cmake/boost_thread-1.70.0/boost_thread-config.cmake

  but it set boost_thread_FOUND to FALSE so package "boost_thread" is
  considered to be NOT FOUND.  Reason given by package:

  No suitable build variant has been found.

Call Stack (most recent call first):
  /usr/local/lib/cmake/Boost-1.70.0/BoostConfig.cmake:124 (boost_find_dependency)
  /usr/local/Cellar/cmake/3.14.2/share/cmake/Modules/FindBoost.cmake:266 (find_package)
  /usr/local/Cellar/cgal/4.14/lib/cmake/CGAL/CGAL_SetupBoost.cmake:48 (find_package)
  /usr/local/Cellar/cgal/4.14/lib/cmake/CGAL/CGALConfig.cmake:87 (include)
  /usr/local/Cellar/cgal/4.14/lib/cmake/CGAL/CGALConfig.cmake:142 (check_cgal_component)
  CMakeLists.txt:2 (find_package)
```

**Comment by scpeters** [comment]
This is weird; why is `osm2pgrouting` being built before `sfcgal`? I figured the topological sort would put it after `sfcgal`? Is this a bug in `test-bot`? (The fact that `sfcgal` is failing to build from source is a separate issue with finding boost thread).

~~~
22:01:02 ==> brew install --only-dependencies osm2pgrouting
22:01:26 ==> FAILED
22:01:26 ==> Installing dependencies for osm2pgrouting: libpqxx, protobuf-c, sfcgal, postgis and pgrouting
22:01:26 ==> Installing osm2pgrouting dependency: libpqxx
22:01:26 ==> Downloading https://homebrew.bintray.com/bottles/libpqxx-6.2.5_5.mojave.bottle.tar.gz
22:01:26 Already downloaded: /Users/brew/Library/Caches/Homebrew/downloads/b13d7d3b39f08ef749edcf5fd0fcd6f808a7e2be765fbcec955c7dffcf8438fb--libpqxx-6.2.5_5.mojave.bottle.tar.gz
22:01:26 ==> Pouring libpqxx-6.2.5_5.mojave.bottle.tar.gz
22:01:26 /usr/local/Cellar/libpqxx/6.2.5_5: 102 files, 1.3MB
22:01:26 ==> Installing osm2pgrouting dependency: protobuf-c
22:01:26 ==> Downloading https://homebrew.bintray.com/bottles/protobuf-c-1.3.1_2.mojave.bottle.tar.gz
22:01:26 Already downloaded: /Users/brew/Library/Caches/Homebrew/downloads/50a06cc3952df301cff9bb33bc2ca3a5a4a3e2102876102420a5e6b74d01829a--protobuf-c-1.3.1_2.mojave.bottle.tar.gz
22:01:26 ==> Pouring protobuf-c-1.3.1_2.mojave.bottle.tar.gz
22:01:26 /usr/local/Cellar/protobuf-c/1.3.1_2: 14 files, 297.0KB
22:01:26 ==> Installing osm2pgrouting dependency: sfcgal
22:01:26 ==> Downloading https://homebrew.bintray.com/bottles/sfcgal-1.3.6_2.mojave.bottle.tar.gz
22:01:26 curl: (22) The requested URL returned error: 404 Not Found
22:01:26 Error: Failed to download resource "sfcgal"
22:01:26 Download failed: https://homebrew.bintray.com/bottles/sfcgal-1.3.6_2.mojave.bottle.tar.gz
22:01:26 Warning: Bottle installation failed: building from source.
~~~

**Comment by scpeters** [comment]
I think a similar thing happened to `pdal`, which built before its dependency `pcl`, whose bottle could not be found. I'll see if I can make a toy example in a tap that illustrates the problem.

**Comment by fxcoudert** [comment]
@scpeters the fact that they build out of order is a missed optimisation, but not in itself a problem. The dependencies will then build from source, and while it will take longer, it should still work.

**Comment by fxcoudert** [comment]
The most serious problem is that many packages using `cmake` can't find boost anymore. This is the same thing that is happening with the simple `cgal` test. With boost 1.69.0, `cmake` shows this:

```
-- Boost version: 1.69.0
-- Found the following Boost libraries:
--   thread
--   system
--   chrono
--   date_time
--   atomic
-- Boost include dirs: /usr/local/include
-- Boost libraries:    /usr/local/lib/libboost_thread-mt.dylib;/usr/local/lib/libboost_system-mt.dylib;/usr/local/lib/libboost_chrono-mt.dylib;/usr/local/lib/libboost_date_time-mt.dylib;/usr/local/lib/libboost_atomic-mt.dylib
```

but with boost-1.70.0, we get this:

```
-- Found Boost 1.70.0 at /usr/local/lib/cmake/Boost-1.70.0
--   Requested configuration: QUIET REQUIRED COMPONENTS thread;system
-- Found boost_headers 1.70.0 at /usr/local/lib/cmake/boost_headers-1.70.0
-- Found boost_thread 1.70.0 at /usr/local/lib/cmake/boost_thread-1.70.0
-- No suitable boost_thread variant has been identified!
--   libboost_thread-mt.dylib (shared, BUILD_SHARED_LIBS not ON, set Boost_USE_STATIC_LIBS=OFF to override)
--   libboost_thread-mt.a (static, Boost_USE_STATIC_LIBS=OFF)
CMake Error at /usr/local/lib/cmake/Boost-1.70.0/BoostConfig.cmake:95 (find_package):
  Found package configuration file:

    /usr/local/lib/cmake/boost_thread-1.70.0/boost_thread-config.cmake

  but it set boost_thread_FOUND to FALSE so package "boost_thread" is
  considered to be NOT FOUND.  Reason given by package:

  No suitable build variant has been found.
```

**Comment by fxcoudert** [comment]
Trying to rebuild `cmake` to see if it can detect `boost` libraries better.

**Comment by fxcoudert** [comment]
- `pcl` failure, fixed upstream: https://github.com/PointCloudLibrary/pcl/pull/2942

**Comment by fxcoudert** [comment]
@BrewTestBot test this please

**Comment by fxcoudert** [comment]
- `libtorrent-rasterbar` fails: https://github.com/arvidn/libtorrent/issues/3683

**Comment by scpeters** [comment]
> the fact that they build out of order is a missed optimisation, but not in itself a problem. The dependencies will then build from source, and while it will take longer, it should still work.

Thanks for the reminder about this behavior. I guess it may happen very often that it fails to find bottles and rebuilds from source, but I only see the console message when it fails, so it can be misleading.

**Comment by Ponce** [comment]
regarding the error

```
No suitable build variant has been found.
```

I'll paste here a relevant comment from Patrick Volkerding in the Slackware ChangeLog

```
l/boost-1.70.0-x86_64-1.txz:  Upgraded.
  Shared library .so-version bump.
  Note: Boost now provides its own BoostConfig.cmake config file, and it may
  not work with all existing code (here, calligra stumbled over it). At this
  point it's not clear if the included cmake config files are buggy, or if
  affected projects need to change something in order to use them, but there's
  an easy workaround to use cmake's FindBoost.cmake (as was used previously).
  Add this to the call to cmake from any affected project (if cmake fails with
  an error: "No suitable build variant has been found."):
  -DBoost_NO_BOOST_CMAKE=ON
```

**Comment by fxcoudert** [comment]
I added `-DBoost_NO_BOOST_CMAKE=ON` to `cgal` but it's not enough. It needs to be added to every project that uses `cgal` apparently…

**Comment by lrineau** [comment]
[**@fxcoudert**](https://github.com/fxcoudert) commented on [Apr 13, 2019, 12:22 PM GMT+2](https://github.com/Homebrew/homebrew-core/pull/38890#issuecomment-482796746 "2019-04-13T10:22:21Z - Replied by Github Reply Comments"):
> Trying to rebuild `cmake` to see if it can detect `boost` libraries better.
> 
> _Edit: didn't work. But, upon further analysis, all packages that fail with this error are CGAL-related, so I filed it as a CGAL issue [CGAL/cgal#3851](https://github.com/CGAL/cgal/issues/3851)_

As far as I can see in the discussion, there are other packages, not related to CGAL, that fail to build.


[**@fxcoudert**](https://github.com/fxcoudert) commented on [Apr 15, 2019, 1:11 PM GMT+2](https://github.com/Homebrew/homebrew-core/pull/38890#issuecomment-483209344 "2019-04-15T11:11:56Z - Replied by Github Reply Comments"):

>     -- Found Boost 1.70.0 at /usr/local/lib/cmake/Boost-1.70.0
>     --   Requested configuration: QUIET REQUIRED COMPONENTS thread;system;regex;date_time;program_options;filesystem
>     -- Found boost_headers 1.70.0 at /usr/local/lib/cmake/boost_headers-1.70.0
>     -- Found boost_thread 1.70.0 at /usr/local/lib/cmake/boost_thread-1.70.0
>     -- No suitable boost_thread variant has been identified!
>     --   libboost_thread-mt.dylib (shared, BUILD_SHARED_LIBS not ON, set Boost_USE_STATIC_LIBS=OFF to override)
>     --   libboost_thread-mt.a (static, Boost_USE_STATIC_LIBS=OFF)
>     CMake Error at /usr/local/lib/cmake/Boost-1.70.0/BoostConfig.cmake:95 (find_package):
>       Found package configuration file:
> 
>         /usr/local/lib/cmake/boost_thread-1.70.0/boost_thread-config.cmake
> 
>       but it set boost_thread_FOUND to FALSE so package "boost_thread" is
>       considered to be NOT FOUND.  Reason given by package:
> 
>       No suitable build variant has been found.

It seems to be an incoherence between `BUILD_SHARED_LIBS` and `Boost_USE_STATIC_LIBS`. Do you know why "`BUILD_SHARED_LIBS` not `ON`"?

On Linux, with Brew, `BUILD_SHARED_LIBS` seems `ON` by default. If I force it to `OFF`, than the formula fails to build as well, with the same error messages, unless I set `CGAL_Boost_USE_STATIC_LIBS=ON` also.

**Comment by fxcoudert** [comment]
> As far as I can see in the discussion, there are other packages, not related to CGAL, that fail to build

Yes, but usually with different errors 😢 

> Do you know why "BUILD_SHARED_LIBS not ON"?

No, it is not something we (Homebrew) are setting.

**Comment by scpeters** [comment]
I've been looking at the cgal issues a little, but so far it's just been tinkering with their travis scripts to try to enable macOS CI to make it easier to reproduce these failures. No solutions yet.

**Comment by fxcoudert** [comment]
Trying to adapt the patch from https://github.com/CGAL/cgal/pull/3886 as it does not apply cleanly

**Comment by fxcoudert** [comment]
Trying an alternative route: compiling boost with `--no-cmake-config`

**Comment by fxcoudert** [comment]
If we ask boost not to install cmake headers, we have…

- more cgal-related failures @lrineau :( 

```
/tmp/sfcgal-20190509-30832-1cn3xqs/SFCGAL-1.3.6/src/Coordinate.cpp:276:32: error: no matching function for call to 'round'
                               SFCGAL::round( v.exact() * _scaleFactor ),
                               ^~~~~~~~~~~~~
/tmp/sfcgal-20190509-30832-1cn3xqs/SFCGAL-1.3.6/include/SFCGAL/numeric.h:50:15: note: candidate function not viable: no known conversion from 'typename enable_if<is_compatible_arithmetic_type<long, number<gmp_rational, et_on> >, detail::expression<detail::multiply_immediates, number<gmp_rational, et_on>, long> >::type' (aka 'boost::multiprecision::detail::expression<boost::multiprecision::detail::multiply_immediates, boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, boost::multiprecision::et_on>, long, void, void>') to 'const double' for 1st argument
inline double round( const double& v )
              ^
/tmp/sfcgal-20190509-30832-1cn3xqs/SFCGAL-1.3.6/include/SFCGAL/numeric.h:86:23: note: candidate function not viable: no known conversion from 'typename enable_if<is_compatible_arithmetic_type<long, number<gmp_rational, et_on> >, detail::expression<detail::multiply_immediates, number<gmp_rational, et_on>, long> >::type' (aka 'boost::multiprecision::detail::expression<boost::multiprecision::detail::multiply_immediates, boost::multiprecision::number<boost::multiprecision::backends::gmp_rational, boost::multiprecision::et_on>, long, void, void>') to 'const CGAL::Gmpq' for 1st argument
SFCGAL_API CGAL::Gmpz round( const CGAL::Gmpq& v ) ;
                      ^
```

- `widelands` needs revision
- `folly` fails with:

```
/tmp/folly-20190509-9923-15qwjok/folly-2019.03.18.00/folly/compression/Zstd.cpp:281:20: error: use of undeclared identifier 'ZSTD_CCtxParam_setParameter'; did you mean 'ZSTD_CCtxParams_setParameter'?
  zstdThrowIfError(ZSTD_CCtxParam_setParameter(params_.get(), param, value));
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
                   ZSTD_CCtxParams_setParameter
/usr/local/include/zstd.h:1463:20: note: 'ZSTD_CCtxParams_setParameter' declared here
ZSTDLIB_API size_t ZSTD_CCtxParams_setParameter(ZSTD_CCtx_params* params, ZSTD_cParameter param, int value);
                   ^
```

**Comment by fxcoudert** [comment]
SFCGAL requires multiple patches and tweaking of compiler flags, and does not appear to be regularly updated to stay compatibly with CGAL. So, I'm removing it.

Within core, it's only use to provide optional functionality (advanced functions) for `postgis`.

**Comment by zbeekman** [comment]
My bad on #40286. I think I should revert that. We may have to delete that bottle from bintray before merging this..

**Comment by chenrui333** [comment]
I can try test this formula in my local?

**Comment by zbeekman** [comment]
> I can try test this formula in my local?

No, please don't worry about this, and let the other maintainers weigh in and help me figure out the right way to get out of the mess I've created. Thanks!

File: Formula/boost-python3.rb
Comment (zbeekman) [inline Formula/boost-python3.rb]: Due to my mistake at #40286 (and there already being a bottle for it on bintray) this likely needs a revision bump. See also #40303.

```suggestion
  sha256 "430ae8354789de4fd19ee52f3b1f739e1fba576f0aded0897c3c2bc00fb38778"
  revision 1
```

Sincerest apologies once again.

**Review by zbeekman — CHANGES_REQUESTED** [review]
Might need this change to fix my mistake.

**Comment by zbeekman** [comment]
I think my revert was the correct move. I left the bottles in place though. Did a bunch of local testing and CI seems to be behaving correctly.

**Comment by SMillerDev** [comment]
Anything I can do to help here @fxcoudert @lembacon ?

**Comment by fxcoudert** [comment]
- SFCGAL have updated their patch, I'm giving it a try.
- Added a revision for `boost-python3` after the mishap & revert

**Comment by scpeters** [comment]
I think avro-cpp might need a revision bump

**Comment by scpeters** [comment]
gnuradio patch in https://github.com/gnuradio/gnuradio/pull/2451

File: Formula/monero.rb
Comment (scpeters) [inline Formula/monero.rb]: this patch doesn't apply

Comment (scpeters) [reply Formula/monero.rb]: it probably needs this monstrosity of over 4k lines to apply: https://github.com/monero-project/monero/commit/973403bc9f54ab0722b67a3c76ab6e7bafbfeedc

**Comment by scpeters** [comment]
the libpulsar 2.3.1 tarball wasn't being found on the apache servers, but 2.3.2 is out so I updated it in #40591

**Comment by scpeters** [comment]
tracking issue for cpprestsdk problems with boost 1.70: https://github.com/microsoft/cpprestsdk/issues/1141

**Comment by scpeters** [comment]
> gnuradio patch in gnuradio/gnuradio#2451

that patch wasn't merged, but the patch from https://github.com/gnuradio/gnuradio/pull/2500 was merged: https://github.com/gnuradio/gnuradio/pull/2500/commits/6dc8229fd0dda25c054c2194ee2c9b28affe92d8

**Comment by scpeters** [comment]
I rebased to resolve the folly and libpulsar conflicts and added a patch for gnuradio; we'll see how goes this time

**Comment by scpeters** [comment]
looks like cpprestsdk includes websocketpp as a git submodule, which is where the boost asio issues are coming from

* discussion: https://github.com/microsoft/cpprestsdk/issues/1141#issuecomment-498002166
* potential fix: https://github.com/zaphoyd/websocketpp/pull/814

**Comment by scpeters** [comment]
reported https://github.com/cruppstahl/upscaledb/issues/123, https://github.com/apache/pulsar/issues/4449

**Comment by scpeters** [comment]
patch for libpulsar: https://github.com/apache/pulsar/commit/07845c5b463b35824f7b4bcab526e90e53489cdb

**Comment by fxcoudert** [comment]
SCFGAL update in progress: https://github.com/Homebrew/homebrew-core/pull/40677 should remove the need for patching this formula

**Comment by scpeters** [comment]
I removed patches from folly and sfcgal and added a patch for libpulsar

**Comment by Al2Klimov** [comment]
FYI:

```
$ brew upgrade boost
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 1 tap (homebrew/core).
No changes to formulae.

==> Upgrading 1 outdated package:
boost 1.69.0_2 -> 1.70.0
==> Upgrading boost
==> Downloading https://homebrew.bintray.com/bottles/boost-1.70.0.mojave.bottle.tar.gz

curl: (22) The requested URL returned error: 401 Unauthorized
Error: Failed to download resource "boost"
Download failed: https://homebrew.bintray.com/bottles/boost-1.70.0.mojave.bottle.tar.gz
Warning: Bottle installation failed: building from source.
==> Downloading https://dl.bintray.com/boostorg/release/1.70.0/source/boost_1_70_0.tar.bz2
```

**Comment by scpeters** [comment]
@Al2Klimov this pull request hasn't been merged yet, so it doesn't have any published bottles. I'm not sure what you're trying to say.

**Comment by fxcoudert** [comment]
We're closer than ever! Remaining issues:
- `cpprestsdk` fails with
```/usr/local/include/boost/asio/impl/executor.hpp:179:22: error: no member named 'context' in 'std::__1::reference_wrapper<boost::asio::io_context>'
```
- `monero` patch does not apply

**Comment by fxcoudert** [comment]
I've manually edited patches for monero and cpprestsdk based on upstream. Not sure if they will work, but cannot test locally right now.

**Comment by fxcoudert** [comment]
Woohoo finally merged! Many many thanks to @scpeters

**Comment by rkooo567** [comment]
I faced this error while building my program. I guess it can be related to this issue? I am using boost 1.70.0 downloaded by brew on MacOS High Sierra 10.13.6. (I used `brew install boost`)

```
/Users/sangbin.cho/classes/riselab/logging_project/clipper/src/libs/httpserver/server_http.hpp:302:84: error: no member named 'get_io_service' in 'boost::asio::basic_stream_socket<boost::asio::ip::tcp,
      boost::asio::executor>'
        timer = std::unique_ptr<asio::steady_timer>(new asio::steady_timer(socket->get_io_service()));
                                                                           ~~~~~~  ^
/Users/sangbin.cho/classes/riselab/logging_project/clipper/src/libs/httpserver/server_http.hpp:638:28: note: in instantiation of member function
      'SimpleWeb::ServerBase<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor> >::Connection::set_timeout' requested here
      session->connection->set_timeout(config.timeout_request);
                           ^
/Users/sangbin.cho/classes/riselab/logging_project/clipper/src/libs/httpserver/server_http.hpp:933:17: note: in instantiation of member function
      'SimpleWeb::ServerBase<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor> >::read' requested here
          this->read(session);
                ^
/Users/sangbin.cho/classes/riselab/logging_project/clipper/src/libs/httpserver/server_http.hpp:72:207: error: no member named 'get_io_service' in 'boost::asio::basic_stream_socket<boost::asio::ip::tcp,
      boost::asio::executor>'
  ...session_, long timeout_content) noexcept : std::ostream(nullptr), session(std::move(session_)), timeout_content(timeout_content), strand(session->connection->socket->get_io_service()) {
                                                                                                                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~  ^
/Users/sangbin.cho/classes/riselab/logging_project/clipper/src/libs/httpserver/server_http.hpp:646:57: note: in instantiation of member function
      'SimpleWeb::ServerBase<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor> >::Response::Response' requested here
          auto response = std::shared_ptr<Response>(new Response(session, this->config.timeout_content));
                                                        ^
/Users/sangbin.cho/classes/riselab/logging_project/clipper/src/libs/httpserver/server_http.hpp:933:17: note: in instantiation of member function
      'SimpleWeb::ServerBase<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor> >::read' requested here
          this->read(session);
```

**Comment by fxcoudert** [comment]
@rbooo567 yes you need to make your program compatible with boost's changes in asio

**Comment by dnsmichi** [comment]
@fxcoudert A slight misunderstanding by @Al2Klimov how source and bottles work, all things sorted out, thanks. We're developing icinga2 on our Macbooks and using Homebrew's Boost for this. 1.70 and its breaking changes came to our attention with FreeBSD's updates, and using Homebrew is just an awesome possibility :)

**Pull request closed without merging**
